File: 0225-implement-stack-using-queues/0225-implement-stack-using-queues.cpp
class MyStack {
public:
    queue<int> q1, q2;
    int cnt;
    MyStack() {
        cnt = 0;
    }
    
    void push(int x) {
        q1.push(x);
        cnt++;
    }
    
    int pop() {
        for(int i = 0; i < cnt - 1; i++){
            q2.push(q1.front());
            q1.pop();
        }
        int ans = q1.front();
        q1.pop();
        while(!q2.empty()){
            q1.push(q2.front());
            q2.pop();
        }
        cnt--;
        return ans;
    }
    
    int top() {
        return q1.back();
    }
    
    bool empty() {
        return q1.empty();
        //return !cnt;
    }

    //  https://leetcode.com/problems/implement-stack-using-queues/solutions/3970478/best-solution-on-leetcode/
    //  이 방법은 stack의 top 즉 가장 마지막에 들어오는 원소를 queue의 가장 앞에 오게 하도록
    //  q1에 있던 원소를 q2로 옮기고 q1에 x를 추가한 후 q2의 원소를 다시 q1으로 옮긴다
    //  
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */

File: 0225-implement-stack-using-queues/0225-implement-stack-using-queues.cs
public class MyStack {
    //  https://leetcode.com/problems/implement-stack-using-queues/solutions/3969470/99-74-one-queue-approach/
    //  c++ 주석에 있는 방법과 같은 원리
    //  push 할 때 x를 push 하고 que의 개수 - 1 만큼 Dequeue한 원소를 다시 Enqueue해서 가장 마지막에 
    //  추가한 x가 제일 앞에 오게함
    Queue<int> que;
    public MyStack() {
        que = new Queue<int>();
    }
    
    public void Push(int x) {
        que.Enqueue(x);
        for(int i = 0; i < que.Count - 1; i++){
            que.Enqueue(que.Dequeue());
        }
    }
    
    public int Pop() {
        return que.Dequeue();
    }
    
    public int Top() {
        return que.Peek();
    }
    
    public bool Empty() {
        return que.Count == 0;
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 * int param_3 = obj.Top();
 * bool param_4 = obj.Empty();
 */